app/services/user_role_service.rb

```diff
@@ -30,6 +30,20 @@ def update_tag_moderators(user_ids, tag)
     true
   end
 
+  def create_or_update_note(reason, content)
+    note = Note.find_by(noteable_id: @user.id, noteable_type: "User", reason: reason)
+    if note.nil?
+      Note.create(
+        noteable_id: @user.id,
+        noteable_type: "User",
+        reason: reason,
+        content: content,
+      )
+    else
+      note.update(content: content)
+    end
```

Comment:
```
note = Note.find_or_create_by(noteable_id: @user.id, noteable_type: "User", reason: reason)
note.update(content: content)
```

This has a slight performance hit because you are doing two write calls in case the Note doesn't exist vs doing just one with the `Note.create`, but it's easier to read.

Reply:
Is it though? The `if` is pretty clear in my opinion.

I'm sure than in a few weeks or months if someone would have to change that code would wonder: wait, why are we doing two writes? 

At least I would :)

Reply:
BTW this would probably be a oneliner with PostgreSQL's "upsert" if Rails supported it

Reply:
@rhymes I think the `nil?` is what made me do a double check, if it said `note.exists?` it'd be easier to parse at glance but `find_by` doesn't support it :/ `find_or_create_by` would also be easier to maintain as you add more fields to Note because you don't have to update two parameters hashes. All in all both solutions are good, it's just a personal preference. I think it's always good to propose different approaches because discussing them might surface new ideas :)

There's actaully [a gem for adding upsert support to AR](https://github.com/jesjos/active_record_upsert) but I haven't used it personally. Did you try it anywhere?

Reply:
You're right, what about this version?

```ruby
def create_or_update_note(reason, content)
  note = Note.find_by(noteable_id: @user.id, noteable_type: "User", reason: reason)
  if note.present?
    note.update(content: content)
  else
    Note.create(
      noteable_id: @user.id,
      noteable_type: "User",
      reason: reason,
      content: content,
    )
  end
end
```

I just switched `nil?` with `present?` so that the first condition you encounter is the update. 

ps. no, I haven't tried UPSERT yet